<p style="color:white; background:#07D; padding:12px; font-size:20px; font-style:italic; text-align:center">
<span style="width:49%; display:inline-block; text-align:left">Christophe Schlick</span>
<span style="width:49%; display:inline-block; text-align:right">schlick[at]u-bordeaux.fr</span>
<span style="font-size:48px; font-style:normal"><b>EXERCICES : PANDAS</b></span><br>
<span style="width:49%; display:inline-block; text-align:left">Version 2022-11</span>
<span style="width:49%; display:inline-block; text-align:right">Licence CC-BY-NC-ND</span></p>

L'objectif de cette feuille d'exercices est de voir comment mettre en oeuvre quelques unes des fonctionnalités fournies par le package **pandas**, basées sur l'utilisation des conteneurs **`series`** et **`dataframe`**.
En parallèle, pour l'affichage des données, on mettra en oeuvre la bibliothèque **seaborn**, une sur-couche de **matplotlib** spécialisée pour les figures liées à l'analyse statistiques des données. Avant de commencer cette feuille d'exercices, il est donc fortement recommandé de relire le notebook **07-Pandas.ipynb** qui contient les éléments de cours correspondants.

In [1]:
import numpy as np
import matplotlib.pyplot as mp
import pandas as pd # import du package 'pandas' avec alias 'pd'
import seaborn as sb # import du package 'seaborn' avec alias 'sb'
sb.set_style('ticks') # on choisit le style des figures (autres styles = white, dark, whitegrid, darkgrid)

<h2 style="padding:16px; color:white; background:#07D">Exercice F1 - Pokemon</h2>

On fournit un fichier CSV [**pokemon.csv**](https://www.labri.fr/perso/schlick/outinfo/CSV/pokemon.csv) contenant des informations diverses sur les 151 Pokemon de première génération.

En ouvrant le fichier directement avec l'environnement Jupyter, on peut constater que la première ligne du fichier contient les noms des colonnes de la table, et que la première colonne (nommée **#**) contient l'indice numérique pour les lignes. Ecrire les instructions permettant de récupérer le contenu du fichier sous la forme d'une table **pandas** en utilisant la colonne **#** pour l'index des lignes, et les labels fournis pour l'index des colonnes :

In [2]:
pkm = 'TODO'

In [3]:
pkm = pd.read_csv('CSV/pokemon.csv') # lecture du fichier (par défaut, la première ligne sert d'index)
pkm.index = pkm['#']; pkm.index.name = None; del pkm['#'] # on transforme la colonne '#' en index
pkm

,Name,Type 1,Type 2,Total,HP,Atk,Def,Spec Atk,Spec Def,Speed,Stage,Legendary
1,Bulbasaur,Grass,Poison,318,45,49,49,65,65,45,1,False
2,Ivysaur,Grass,Poison,405,60,62,63,80,80,60,2,False
3,Venusaur,Grass,Poison,525,80,82,83,100,100,80,3,False
4,Charmander,Fire,NaN,309,39,52,43,60,50,65,1,False
5,Charmeleon,Fire,NaN,405,58,64,58,80,65,80,2,False
...,...,...,...,...,...,...,...,...,...,...,...,...
147,Dratini,Dragon,NaN,300,41,64,45,50,50,50,1,False
148,Dragonair,Dragon,NaN,420,61,84,65,70,70,70,2,False
149,Dragonite,Dragon,Flying,600,91,134,95,100,100,80,3,False
150,Mewtwo,Psychic,NaN,680,106,110,90,154,90,130,1,True


---
Les labels de certaines colonnes comportent des espaces, ce qui interdit de les utiliser avec un accès par attribut (notation pointée). Ecrire les instructions permettant de supprimer ces espaces :

In [4]:
# TODO

In [5]:
pkm.columns = [name.replace(' ','') for name in pkm.columns]
pkm

,Name,Type1,Type2,Total,HP,Atk,Def,SpecAtk,SpecDef,Speed,Stage,Legendary
1,Bulbasaur,Grass,Poison,318,45,49,49,65,65,45,1,False
2,Ivysaur,Grass,Poison,405,60,62,63,80,80,60,2,False
3,Venusaur,Grass,Poison,525,80,82,83,100,100,80,3,False
4,Charmander,Fire,NaN,309,39,52,43,60,50,65,1,False
5,Charmeleon,Fire,NaN,405,58,64,58,80,65,80,2,False
...,...,...,...,...,...,...,...,...,...,...,...,...
147,Dratini,Dragon,NaN,300,41,64,45,50,50,50,1,False
148,Dragonair,Dragon,NaN,420,61,84,65,70,70,70,2,False
149,Dragonite,Dragon,Flying,600,91,134,95,100,100,80,3,False
150,Mewtwo,Psychic,NaN,680,106,110,90,154,90,130,1,True


---
Vérifier les types de données pour les 12 colonnes :

In [6]:
# TODO

In [7]:
pkm.dtypes # type de données par colonne

Name         object
Type1        object
Type2        object
Total         int64
HP            int64
Atk           int64
Def           int64
SpecAtk       int64
SpecDef       int64
Speed         int64
Stage         int64
Legendary      bool
dtype: object

Afficher les statistiques descriptives pour les colonnes numériques de la table :

In [8]:
# TODO

In [9]:
pkm.describe() # statistiques descriptives par colonne

,Total,HP,Atk,Def,SpecAtk,SpecDef,Speed,Stage
count,151.00000,151.000000,151.000000,151.000000,151.000000,151.000000,151.000000,151.000000
mean,407.07947,64.211921,72.549669,68.225166,67.139073,66.019868,68.933775,1.582781
std,99.74384,28.590117,26.596162,26.916704,28.534199,24.197926,26.746880,0.676832
min,195.00000,10.000000,5.000000,5.000000,15.000000,20.000000,15.000000,1.000000
25%,320.00000,45.000000,51.000000,50.000000,45.000000,49.000000,46.500000,1.000000
50%,405.00000,60.000000,70.000000,65.000000,65.000000,65.000000,70.000000,1.000000
75%,490.00000,80.000000,90.000000,84.000000,87.500000,80.000000,90.000000,2.000000
max,680.00000,250.000000,134.000000,180.000000,154.000000,125.000000,140.000000,3.000000


Afficher le nombre de valeurs distinctes pour chaque colonne :

In [10]:
# TODO

In [11]:
pkm.nunique() # nombre de valeurs distinctes par colonne

Name         150
Type1         15
Type2         11
Total         74
HP            42
Atk           46
Def           44
SpecAtk       34
SpecDef       29
Speed         42
Stage          3
Legendary      2
dtype: int64

---
On constate qu'il y a 15 noms différents pour **Type1** et 11 pour **Type2**. Ecrire les instructions permettant de les afficher, classés par ordre alphabétique : 

In [12]:
# TODO

In [13]:
combine = lambda index: f"{index} : {' '.join(sorted(pkm[index].dropna().unique()))}" # on factorise
print(combine('Type1')) # on applique la fonction 'combine' à la colonne 'Type1'
print(combine('Type2')) # idem pour 'Type2'

Type1 : Bug Dragon Electric Fairy Fighting Fire Ghost Grass Ground Ice Normal Poison Psychic Rock Water
Type2 : Fairy Fighting Flying Grass Ground Ice Poison Psychic Rock Steel Water


On constate également qu'il y a 150 noms différents alors qu'il y a 151 lignes. Trouver le nom en doublon :

In [14]:
# TODO

In [15]:
print(pkm.Name[pkm.Name.duplicated()]) # on trouve 'Nidoran'
pkm[pkm.Name == 'Nidoran'] # on affiche les 2 lignes concernées

32    Nidoran
Name: Name, dtype: object


,Name,Type1,Type2,Total,HP,Atk,Def,SpecAtk,SpecDef,Speed,Stage,Legendary
29,Nidoran,Poison,NaN,275,55,47,52,40,40,41,1,False
32,Nidoran,Poison,NaN,273,46,57,40,40,40,50,1,False


Trouver sur le web la raison de ce doublon, et en déduire un moyen de le corriger dans la table :

In [16]:
# TODO

In [17]:
# 'Nidoran' est le seul Pokemon sexué de la génération 1 (cf. liens ci-dessous)
# https://www.pokebip.com/pokedex/pokemon/nidoran-m  ou  https://www.pokepedia.fr/Nidoran%E2%99%82
# https://www.pokebip.com/pokedex/pokemon/nidoran-f  ou  https://www.pokepedia.fr/Nidoran%E2%99%80
pkm.loc[29,'Name'] = 'NidoranF' # version femelle
pkm.loc[32,'Name'] = 'NidoranM' # version mâle
pkm[28:34]

,Name,Type1,Type2,Total,HP,Atk,Def,SpecAtk,SpecDef,Speed,Stage,Legendary
29,NidoranF,Poison,NaN,275,55,47,52,40,40,41,1,False
30,Nidorina,Poison,NaN,365,70,62,67,55,55,56,2,False
31,Nidoqueen,Poison,Ground,505,90,92,87,75,85,76,3,False
32,NidoranM,Poison,NaN,273,46,57,40,40,40,50,1,False
33,Nidorino,Poison,NaN,365,61,72,57,55,55,65,2,False
34,Nidoking,Poison,Ground,505,81,102,77,85,75,85,3,False


---
La colonne **Total** semble être égale à la somme des colonnes de **HP** à **Speed**. Confirmer ou infirmer cette hypothèse :

In [ ]:
# TODO

En utilisant la fonction **`displot`** du package **seaborn**, afficher l'histogramme de la colonne **Total** ainsi que la courbe d'estimation de sa densité de probabilité (***kde*** = kernel density estimation) :

In [ ]:
# TODO

En utilisant la méthode **`groupby`**, écrire les instructions permettant de calculer la valeur moyenne de la colonne **Total** pour chacune des 12 catégories définies par **Type1** (afficher le résultat par ordre décroissant des moyennes obtenues) :

In [ ]:
# TODO

En utilisant les fonctions **`violinplot`** et **`swarmplot`** de **seaborn**, afficher les variations de la colonne **Total** pour chacun des Pokemon en fonction de son **Type1**, puis de son **Type2** :

In [ ]:
# TODO

---
On s'intéresse maintenant aux 4 caractéristiques **Atk**, **SpecAtk**, **Def** et **SpecDef**. Visualiser les variations de ces 4 colonnes à l'aide de la fonction **`boxplot`** (boîtes à moustaches) :

In [ ]:
# TODO

On cherche à établir s'il existe des corrélations entre ces 4 paramètres. L'outil idéal pour avoir une vision d'ensemble est la fonction **`pairplot`** :

In [ ]:
# TODO

Les diagrammes ne sont pas très probants, le seul endroit où une correlation semble apparaître concerne les variables **SpecAtk** et **SpecDef** (et encore, c'est uniquement pour la première moitié de l'échelle des valeurs). On va utiliser la fonction **`lmplot`** pour visualiser les relations entre ces 2 variables, et appliquer une régression linéaire pour chacune des 3 classes de la colonne **Stage** :

In [ ]:
# TODO

<h2 style="padding:16px; color:white; background:#07D">Exercice F2 - Fortune 500</h2>

On fournit un fichier CSV [**fortune500.csv**](https://www.labri.fr/perso/schlick/outinfo/CSV/fortune500.csv) qui contient 51 classements annuels (entre 1955 et 2005) établis par le magazine [**Fortune**](https://fortune.com/fortune500) et contenant les noms des 500 plus grandes entreprises américaines (au sens du chiffre d'affaires annuel).

Comme dans l'exercice prédédent, on peut constater que la première ligne du fichier contient les noms des colonnes de la table. Ecrire les instructions permettant de récupérer le contenu du fichier sous la forme d'une table **pandas** en gardant l'index par défaut pour les lignes et en utilisant les labels fournis pour l'index des colonnes. Ensuite supprimer la colonne **`Profit (in millions)`** dont on ne se servira pas, et renommer la colonne **`Revenue (in millions)`** en **`Revenue`** :

In [ ]:
top500 = 'TODO'

---
On veut connaitre le nom de toutes les sociétés qui ont été classsées **Top 1** au moins une fois sur les 51 classements annuels, et les afficher par ordre décroissant du nombre de classements dans cette position : 

In [ ]:
top1 = 'TODO'

Même question avec un classement dans le **Top 10**, en se limitant aux sociétés qui y ont été classées au moins 20 fois :

In [ ]:
top10 = 'TODO'

---
Afficher sur un même graphique avec la fonction **`lineplot`**, l'évolution du chiffre d'affaires moyen du top 1, du top 10 et de l'ensemble du top 500 entre les années 1955 et 2005 :

In [ ]:
revenues = 'TODO'

Le défaut majeur de ce graphique est qu'il n'inclut pas l'effet de l'inflation (**1\$** de 1955 n'a évidemment rien à voir avec **1\$** de 2005). Pour corriger cette erreur, il faut inclure l'évolution annuelle de l'indice des prix à la consommation aux Etats-Unis (***Historical Consumer Price Index***) qui se trouve (par exemple) sur [**cette page**](https://inflationdata.com/Inflation/Consumer_Price_Index/HistoricalCPI.aspx). Il faut donc récupérer cette table (via la fonction **`pd.read_html`**), ne garder que la colonne de décembre et les lignes entre 1955 et 2005, et s'en servir pour regénérer le graphique précédent en dollars constants :

In [ ]:
cpi = 'TODO'

---
Et pour finir, on cherche à afficher l'histogramme empilé des chiffres d'affaires (transformés en dollars constants) des sociétés du **Top 10**, pour l'ensemble des 51 classements. Pour cela il faut d'abord
créer une colonne pour chacun des 10 rangs, à l'aide de la fonction **`pivot_table`** puis utiliser la méthode **`plot`** avec les paramètres **`kind='bar'`** et **`stacked=True`** :

In [ ]:
import warnings; warnings.filterwarnings('ignore') # on enlève les 'warning' pour 'setting with copy'
# TODO

<h2 style="padding:16px; color:white; background:#07D">Exercice F3 - Population des USA</h2>

On cherche à établir quelques statistiques sur (une partie) de la base de données du recensement aux Etats-Unis. On fournit trois fichiers :

* L'évolution de la population, Etat par Etat, entre 1990 et 2013 : [**usa-pops.csv**](https://www.labri.fr/perso/schlick/outinfo/CSV/usa-pops.csv)
* La superficie de chaque Etat (en squared miles) : [**usa-areas.csv**](https://www.labri.fr/perso/schlick/outinfo/CSV/usa-areas.csv)
* L'abréviation et le nom officiel de chacun des 50 Etats : [**usa-names.csv**](https://www.labri.fr/perso/schlick/outinfo/CSV/usa-names.csv)

A partir de ces fichiers, et sans avoir plus d'explications sur la manière de procéder, écrire la liste des instructions à exécuter pour obtenir les résultats suivants :

* Convertir l'ensemble des surfaces des 50 Etats (uniquement les 50 Etats fédéraux) en **km$^2$** et afficher les statistiques descriptives associées
* Vérifier si, pour chaque année, la population totale des USA est bien égale à la somme de celle des 50 Etats, ou si les totaux fournis par le fichier incluent la population des 2 territoires non-fédéraux :<br> **DC** = *District of Columbia* et **PR** = *Puerto Rico*
* Tracer l'évolution de la population globale des Etats-Unis entre 1990 et 2013, selon trois courbes différentes : les personnes mineures, les personnes majeures, la somme des deux
* Calculer l'évolution de la densité de population (habitants par km$^2$) pour chaque Etat, puis tracer (sur une même figure) l'évolution des 3 Etats les plus denses et des 3 Etats les moins denses
* Calculer pour chaque année, l'évolution relative de la population de chaque Etat, puis tracer sous forme
de nuage de points, la relation entre la population totale de chaque Etat et son évolution relative

> **Note :** La table des superficies utilise les noms complets des Etats, alors que la table des populations utilise le sigle sur 2 caractères. Pour utiliser correctement les deux tables, il faudra donc se servir de la troisième table qui associe les noms et les sigles pour les 50 Etats fédéraux.

In [ ]:
# lecture des fichiers et affichage des 2 premières et 2 dernières lignes
names = pd.read_csv('CSV/usa-names.csv'); print(f"names:\n{names.iloc[[0,1,-2,-1],:]}\n")
areas = pd.read_csv('CSV/usa-areas.csv'); print(f"areas:\n{areas.iloc[[0,1,-2,-1],:]}\n")
pops = pd.read_csv('CSV/usa-pops.csv'); print(f"pops:\n{pops.iloc[[0,1,-2,-1],:]}\n")

In [ ]:
# TODO

<h2 style="padding:16px; color:white; background:#07D">Exercice F3 - Equipements publics Bordeaux Metropole</h2>

La métropole de Bordeaux propose un grand nombre de [**jeux de données en OpenData**](https://opendata.bordeaux-metropole.fr) dont on a extrait les trois fichiers suivants (en supprimant quelques colonnes pour alléger la taille des données) :

* La liste (relativement) complète des adresses postales de la métropole : [**bx-adresses.csv**](https://www.labri.fr/perso/schlick/outinfo/CSV/bx-adresses.csv)
* La liste complète des équipements publics, classé en 15 catégories : [**bx-equi-pub.csv**](https://www.labri.fr/perso/schlick/outinfo/CSV/bx-equi-pub.csv)
* La liste des codes et des titres asssociés à ces 15 catégories : [**bx-code-equi.csv**](https://www.labri.fr/perso/schlick/outinfo/CSV/bx-code-equi.csv)

On cherche à exploiter ces fichiers pour obtenir une visualisation géolocalisée des équipements publics de la métropole. Il existe plusieurs bibliothèques pour Python permettant la création et la visualisation de cartes contenant des données géolocalisées, la plus simple d'accès est très certainement la bibliothèque [**folium**](https://python-visualization.github.io/folium) qui offre un "wrapper" simplifié pour la bibliothèque [**leaflet.js**](https://leafletjs.com) écrite en JavaScript. Malheureusement, la bibliothèque **folium** n'est pas installée par défaut avec la distribution **Anaconda**, il faut donc l'installer manuellement pour pouvoir l'utiliser. Cela est possible sur vos postes personnels (les instructions se trouvent dans les liens ci-dessous) mais pas sur les postes du CREMI, où il faudra impérativement passer sur l'environnement **binder** pour utiliser cette bibliothèque.

Pour la mise en oeuvre de **folium**, il est conseillé de commencer par lire les deux (courts) tutos suivants : un [**exemple d'introduction**](https://towardsdatascience.com/creating-a-simple-map-with-folium-and-python-4c083abfff94) tiré du site **towardsdatascience**, et le [**quickstart**](https://python-visualization.github.io/folium/quickstart.html) de la documentation officiel du package.

Ensuite, sans avoir d'explications complémentaires sur la manière de procéder, écrire la liste des instructions à exécuter pour obtenir les résultats suivants :

* Afficher l'histogramme du nombre d'équipements publics pour chacune des 15 catégories
* Créer et afficher une carte contenant l'ensemble des 5600 équipements publics de la métropole : on utilisera un marqueur graphique d'une couleur ou forme différente pour chaque catégorie, et chaque marqueur devra afficher un popup fournissant des informations complètes sur l'équipement sélectionné)
* Pour certains équipement, la donnée de la colonne **`adresse`** est manquante. En utilisant le fichier des adresses et les coordonnées GPS, trouver l'adresse la plus probable pour chacun de ces équipements.


In [ ]:
# lecture des fichiers et affichage des 2 premières et 2 dernières lignes
adres = pd.read_csv('CSV/bx-adresses.csv'); print(f"adres:\n{adres.iloc[[0,1,-2,-1],:]}\n")
equis = pd.read_csv('CSV/bx-equi-pub.csv'); print(f"equis:\n{equis.iloc[[0,1,-2,-1],:]}\n")
codes = pd.read_csv('CSV/bx-code-equi.csv'); print(f"codes:\n{codes.iloc[[0,1,-2,-1],:]}\n")

In [ ]:
# TODO

---